core/src/main/scala/cats/data/OptionT.scala

```diff
@@ -362,6 +362,23 @@ final case class OptionT[F[_], A](value: F[Option[A]]) {
   def filter(p: A => Boolean)(implicit F: Functor[F]): OptionT[F, A] =
     OptionT(F.map(value)(_.filter(p)))
 
+  /**
+   * Example:
+   * {{{
+   *  scala> import cats.data.OptionT
+   *
+   *  scala> val optionT: OptionT[List, Int] = OptionT[List, Int](List(Some(100), None, Some(421), Some(333)))
+   *  scala> optionT.filterF(n => List(n % 100 == 0, n.toString.toSet.size == 1))
+   *
+   *  res0: OptionT[List, Int] = OptionT(List(Some(100), None, None, None, None, None, Some(333)))
+   * }}}
+   */
+  def filterF(p: A => F[Boolean])(implicit F: Monad[F]): OptionT[F, A] =
+    OptionT(F.flatMap(value) {
+      case Some(a) => F.map(p(a))(Option.when(_)(a))
```

Comment:
I'd rather not use `Option.when` since that is call-by.name in the second parameter no?

```scala
case Some(a) => F.map(p(a)) {
  case true => Some(a)
  case false => None
}
case None => F.pure(None)
```

we often try to minimize allocations internally since we have a very high testing bar and want to impose minimal cost on users.

Reply:
I was thinking the call-by-name nature should be intentionally kept because F-value might be unused when condition evaluates to `false`. I used other cats methods for reference - like `getOrElseF` or `FtalMap.ifM` and kept it similar for consistency

Reply:
@johnynek what do you think, does it make sense to reuse an option instance like:
```scala
      case v @ Some(a) =>
        F.map(p(a)) {
          case true  => v
          case false => None
        }
```

Reply:
even better than my code!

Reply:
Pretty the same but a bit shorter (and a bit faster, actually):
```
case v @ Some(a) => F.map(p(a)) { if (_) v else None }
```

Reply:
@satorg done!